README.md
# cryptopals

Matasano Crypto Challenges in clojure.

## License

Copyright © 2017 Electronic Dreams, Inc

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log

Nothing to be seen here.


test/cryptopals/core_test.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges: Test Suite
;;-----------------------------------------------------------------------------

(ns cryptopals.core-test
  (:require [clojure.string :as string]
            [clojure.test :refer :all]
            [cryptopals.utils :refer [bytes->str
                                      bytes->ascii
                                      base64file->bytes]]
            [cryptopals.core :refer :all]))

;;- set 2: challenge 9 --------------------------------------------------------

(deftest pkcs#7-padding-test
  (testing "Implement PKCS#7 padding"
    (is (= "YELLOW SUBMARINE\u0004\u0004\u0004\u0004"
           (bytes->str (pkcs#7-padding 20 "YELLOW SUBMARINE"))))))

;;- set 2: challenge 10 -------------------------------------------------------

(deftest aes-cbc-mode-decrypt-file-test
  (testing "Implement CBC mode: decrypt"
    (let [key "YELLOW SUBMARINE"
          iv (take (count key) (repeat 0))]
      (is (string/starts-with?
            (aes-cbc-mode-decrypt-file key iv "resources/s2c10.txt")
            "I'm back and I'm ringin' the bell \n")))))

(deftest aes-cbc-mode-encrypt-test
  (testing "Implement CBC mode: encrypt"
    (let [key "YELLOW SUBMARINE"
          iv (take (count key) (repeat 0))
          encrypted (base64file->bytes "resources/s2c10.txt")
          decrypted (bytes->ascii (aes-cbc-mode-decrypt key iv encrypted))]
      (aes-cbc-mode-encrypt key iv decrypted))))

;;- set 2: challenge 11 -------------------------------------------------------

(deftest aes-ecb-mode-encrypt-test
  (testing "ECB mode: encrypt"
    (let [key "YELLOW SUBMARINE"
          txt "ABCDEFGHIJKLMNOPQRSTUVWYXYZABCDE"
          txt-bytes (byte-array (map int txt))]
      (is (= (aes-ecb-mode-encrypt key txt)
             (map identity (cryptopals.set1/aes-ecb-mode-cipher :encrypt key txt-bytes)))))))

(deftest encryption-oracle-test
  (testing "An ECB/CBC detection oracle."
    (dotimes [_ 100]
      (is (let [input "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
                [mode encrypted] (encryption-oracle input)
                is-ecb (cryptopals.set1/detect-aes-in-ecb-mode 16 encrypted)]
            (case mode
              :ecb is-ecb
              :cbc (not is-ecb)))))))

;;- set 2: challenge 12 -------------------------------------------------------

(deftest make-oracle-decryptor-test
  (testing "Oracle function that encrypts under ECB using a consistent but unknown key"
    (let [unknown (base64file->bytes "resources/s2c11.txt")
          oracle (make-oracle-decryptor unknown)]
      (testing "Discover the block size of the cipher"
        (is (= 16 (discover-block-size oracle))))
      (testing "Detect ECB mode"
        (let [input "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"]
          (is (cryptopals.set1/detect-aes-in-ecb-mode 16 (oracle input)))))
      (testing "Decryption of hidden text"
        (let [unknown-size (count unknown)
              decryptor (byte-at-a-time-ecb-decryption oracle 16)]
          (is (= (vec unknown)
                 (into [] decryptor (range unknown-size)))))))))


test/cryptopals/set1_test.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges: Set 1: Test Suite
;;-----------------------------------------------------------------------------

(ns cryptopals.set1-test
  (:require [clojure.test :refer :all]
            [cryptopals.set1 :refer :all]))

;;- set 1: challenge 1 --------------------------------------------------------

(deftest hex->base64-test
  (testing "Convert hex to base64"
    (is (= "SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t"
           (hex->base64 "49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d")))))

;;- set 1: challenge 2 --------------------------------------------------------

(deftest fixed-xor-hex-test
  (testing "Fixed XOR"
    (is (= "746865206b696420646f6e277420706c6179"
           (fixed-xor-hex "1c0111001f010100061a024b53535009181c"
                          "686974207468652062756c6c277320657965")))))

;;- set 1: challenge 3 --------------------------------------------------------

(deftest a-z?-test
  (testing "Test for lowercase english characters"
    (is (a-z? \a))
    (is (not (a-z? \A)))
    (is (not (a-z? \?)))))

(deftest letter-frequencies-test
  (testing "English letter frequencies"
    (is (= {\b 3
            \c 4
            \d 6
            \e 12}
          (letter-frequencies "bbb CCCC!dddddd_eeeeeeeeeeee")))))

(deftest non-english-penelty-test
  (testing "Non-English characters penelty"
    (is (= 0.0 (non-english-penelty "abcdefg")))
    (is (= 50.0 (non-english-penelty "._|[]{")))))

(deftest single-byte-xor-cipher-hex-test
  (testing "Single-byte XOR cipher"
    (is (= "Cooking MC's like a pound of bacon"
           (:msg (single-byte-xor-cipher-hex "1b37373331363f78151b7f2b783431333d78397828372d363c78373e783a393b3736"))))))

;;- set 1: challenge 4 --------------------------------------------------------

(deftest detect-single-character-xor-test
  (testing "Detect single-character XOR"
    (is (= "Now that the party is jumping\n"
           (:msg (detect-single-character-xor "resources/s1c4.txt"))))))

;;- set 1: challenge 5 --------------------------------------------------------

(deftest repeating-key-xor-hex-test
  (testing "Implement repeating-key XOR"
    (is (= (str "0b3637272a2b2e63622c2e69692a23693a2a3c6324202d623d63343c2a26226324272765272"
                "a282b2f20430a652e2c652a3124333a653e2b2027630c692b20283165286326302e27282f")
           (repeating-key-xor-hex (str "Burning 'em, if you ain't quick and nimble\n"
                                       "I go crazy when I hear a cymbal")
                                  "ICE")))))

;;- set 1: challenge 6 --------------------------------------------------------

(deftest hamming-distance-test
  (testing "Hamming distance"
    (is (= 37 (hamming-distance (map byte "this is a test")
                                (map byte "wokka wokka!!!"))))))

(deftest break-repeating-key-xor-test
  (testing "Break repeating-key XOR"
    (is (= "Terminator X: Bring the noise"
           (:key (break-repeating-key-xor "resources/s1c6.txt"
                                          {:keysize-min 2
                                           :keysize-max 40
                                           :hamming-blocks 5
                                           :keysize-test-best 3}))))))

;;- set 1: challenge 7 --------------------------------------------------------

(deftest aes-ecb-mode-decrypt-test
  (testing "AES in ECB mode"
    (is (clojure.string/starts-with?
          (aes-ecb-mode-decrypt-file "resources/s1c7.txt" "YELLOW SUBMARINE")
                                     (str "I'm back and I'm ringin' the bell \n"
                                          "A rockin' on the mike while the fly girls yell")))))

;;- set 1: challenge 8 --------------------------------------------------------

(deftest detect-aes-in-ecb-mode-file-test
  (testing "Detect AES in ECB mode"
    (is (= 1 (count (detect-aes-in-ecb-mode-file "resources/s1c8.txt"))))))


test/cryptopals/utils_test.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges: Utility Functions: Test Suite
;;-----------------------------------------------------------------------------

(ns cryptopals.utils-test
  (:require [clojure.test :refer :all]
            [cryptopals.utils :refer :all]))

;;- utils ---------------------------------------------------------------------

(deftest hex->bytes-test
  (testing "Convert hex to byte[]"
    (is (= [0] (map int (hex->bytes "00"))))
    (is (= [-1] (map int (hex->bytes "ff"))))))

(deftest bytes->hex-test
  (testing "Convert byte[] to hex"
    (is (= "00" (bytes->hex (byte-array [0]))))
    (is (= "ff" (bytes->hex (byte-array [-1]))))))

(deftest byte->hex-test
  (testing "Convert int to hex"
    (is (= "00" (byte->hex 0)))
    (is (= "ff" (byte->hex 255)))))

(deftest byte->int-test
  (testing "Convert byte to int"
    (is (= 0 (byte->int (byte 0))))
    (is (= 255 (byte->int (byte -1))))))

(deftest hex->ascii-test
  (testing "Convert hex to ascii string"
    (is (= "ab" (hex->ascii "6162")))))

(deftest ->byte-test
  (testing "Convert number/char byte"
    (is (= 97 (->byte \a)))
    (is (= -1 (->byte -1)))
    (is (= -116 (->byte 140)))
    (is (= 0 (->byte 256)))))

(deftest map-values-test
  (testing "Transforming values in a map"
    (is (= {:a 2 :b 3}
           (map-values inc {:a 1 :b 2})))))

(deftest sqr-err-test
  (testing "Squared error test"
    (is (= 0.0 (sqr-err 5 5)))
    (is (= 0.2 (sqr-err 4 5)))
    (is (= 1.8 (sqr-err 2 5)))
    (is (= 3.2 (sqr-err 1 5)))))

(deftest zip-test
  (testing "Zipping together sequences"
    (is (= [[:a] [:b] [:c]] (zip [:a :b :c])))
    (is (= [[:a :d] [:b :e] [:c :f]] (zip [:a :b :c] [:d :e :f])))
    (is (= [[:a :d :g] [:b :e :h] [:c :f :i]] (zip [:a :b :c] [:d :e :f] [:g :h :i])))
    (is (= [[:a :d :g] [:b :e :h]] (zip [:a :b :c] [:d :e :f] [:g :h])))))


src/cryptopals/set1.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges: Set 1: Basics
;;-----------------------------------------------------------------------------

(ns cryptopals.set1
  (:require [clojure.java.io :as io]
            [clojure.string :as string]
            [cryptopals.utils :refer :all])
  (:import
    [org.apache.commons.codec.binary Base64]
    [javax.crypto Cipher]
    [javax.crypto.spec SecretKeySpec]))

;;- set 1: challenge 1 --------------------------------------------------------

(defn hex->base64
  "Convert string hex to base64."
  [hex]
  (Base64/encodeBase64String (hex->bytes hex)))

;;- set 1: challenge 2 --------------------------------------------------------

(defn fixed-xor
  "Takes two equal-length buffers and produces their XOR combination."
  [x y]
  {:pre [(= (count x) (count y))]}
  (map (partial apply bit-xor) (zip x y)))

(defn fixed-xor-hex
  "Takes two equal-length buffers and produces their XOR combination."
  [hex-x hex-y]
  (let [bytes-x (hex->bytes hex-x)
        bytes-y (hex->bytes hex-y)]
    (bytes->hex (byte-array (fixed-xor bytes-x bytes-y)))))

;;- set 1: challenge 3 --------------------------------------------------------

(defn a-z?
  "Character is between a and z."
  [c]
  (let [x (int c)]
    (and (<= 97 x) (<= x 122))))

(defn letter-frequencies
  "Calculate frequencies of english letters in string."
  [s]
  (let [ls (filter a-z? (string/lower-case s))]
    (frequencies ls)))

(def en-letter-frequencies
  "English language letter frequencies.
   reference: https://en.wikipedia.org/wiki/Letter_frequency"
  {\a 0.08167 \b 0.01492 \c 0.02782 \d 0.04253 \e 0.12702 \f 0.02228 \g 0.02015
   \h 0.06094 \i 0.06966 \j 0.00153 \k 0.00772 \l 0.04025 \m 0.02406 \n 0.06749
   \o 0.07507 \p 0.01929 \q 0.00095 \r 0.05987 \s 0.06327 \t 0.09056 \u 0.02758
   \v 0.00978 \w 0.02360 \x 0.00150 \y 0.01974 \z 0.00074})

(defn non-english-penelty
  "Penalize all characters not within the normal alphabet and punctuation."
  [s]
  (letfn [(bad-char [c]
            (not (re-matches #"[a-zA-Z0-9?!,.'\n\- ]+" (str c))))]
    (reduce #(+ %1 (if (bad-char %2) 10.0 0.0)) 0 s)))

(defn is-english?-score
  "Use chi-squared testing to score a string as english."
  [s]
  (try (let [s-freqs (letter-frequencies s)
             s-len (reduce + (vals s-freqs))
             en-freq-err (fn [k v] (sqr-err (get s-freqs k 0) (* s-len v)))]
         (+ (reduce-kv #(+ %1 (en-freq-err %2 %3)) 0 en-letter-frequencies)
            (non-english-penelty s)))
       (catch Exception e nil)))

(defn single-byte-xor-cipher
  "Find the single key xor and decrypt the message."
  [input]
  (letfn [(repeat-key [k]
            (take (count input) (repeat k)))
          (score-test [s]
            {:msg (apply str s)
             :score (is-english?-score s)})
          (merge-keys [scores test-keys]
            (map (fn [[s k]] (assoc s :key k)) (zip scores test-keys)))]
    (let [xf (comp (map repeat-key)
                   (map (partial fixed-xor input))
                   (map bytes->ascii)
                   (map score-test))
          test-keys (range 0 256)]
      (->> (merge-keys (into [] xf test-keys) test-keys)
           (remove #(nil? (:score %)))
           (sort-by :score)
           first))))

(defn single-byte-xor-cipher-hex
  "A hex encoded string has been XOR'd against a single character.
  Find the key, decrypt the message."
  [hex]
  (single-byte-xor-cipher (hex->bytes hex)))

;;- set 1: challenge 4 --------------------------------------------------------

(defn detect-single-character-xor
  "One of the 60-character strings in the file has been encrypted by
  single-character XOR. Find it."
  [file-path]
  (->> (string/split (slurp file-path) #"\n")
       (map single-byte-xor-cipher-hex)
       (sort-by :score)
       first))

;;- set 1: challenge 5 --------------------------------------------------------

(defn repeating-key-xor
  "Encrypt input, under the key, using repeating-key XOR."
  [input xor-key]
  (letfn [(repeat-xor [xor-key length]
            (let [xf (comp cat
                           (take length))
                  byte-key (map byte xor-key)]
              (into [] xf (repeat byte-key))))]
    (fixed-xor (map byte input)
               (repeat-xor xor-key (count input)))))

(defn repeating-key-xor-hex
  "Encrypt character input, under the key, using repeating-key XOR."
  [text xor-key]
  (bytes->hex (byte-array (repeating-key-xor text xor-key))))

;;- set 1: challenge 6 --------------------------------------------------------

(defn hamming-distance
  [x y]
  "The Hamming distance is just the number of differing bits."
  {:pre [(= (count x) (count y))]}
  (letfn [(bit-diff [[x y]]
            (Integer/bitCount (bit-xor x y)))]
    (reduce + (map bit-diff (zip x y)))))

(defn score-hamming [n file-bytes keysize]
  "Use n blocks of keysize to calculate average hamming distance between each."
  (let [blocks (take n (partition keysize file-bytes))
        scores (for [x blocks y blocks :while (not= x y)]
                 (/ (hamming-distance x y) keysize))]
    {:keysize keysize
     :score (/ (reduce + scores) (count scores))}))

(defn extract-xor-key
  "Create blocks of keysize, transpose and find best single bytekey.  Repeat for
  all blocks for the complete key."
  [file-bytes keysize]
  (let [get-key (comp byte->ascii
                      :key
                      single-byte-xor-cipher)
        transposed (apply zip (partition keysize file-bytes))]
    (apply str (map get-key transposed))))

(defn break-repeating-key-xor
  "Breaking repeating-key XOR (Vigenere) statistically is obviously an academic
  exercise, a 'Crypto 101' thing. But more people 'know how' to break it than
  can actually break it, and a similar technique breaks something much more
  important."
  [file-path {:keys [keysize-min
                     keysize-max
                     hamming-blocks
                     keysize-test-best]}]
  (let [file-bytes (base64file->bytes file-path)
        keysizes (range keysize-min keysize-max)
        get-score (partial score-hamming hamming-blocks file-bytes)
        get-key (partial extract-xor-key file-bytes)
        assoc-key #(assoc % :key (get-key (:keysize %)))
        is-english? #(is-english?-score (map char (repeating-key-xor file-bytes %)))
        assoc-english #(assoc % :is-english (is-english? (:key %)))]
    (->> (map get-score keysizes)
         (sort-by :score)
         (take keysize-test-best)
         (map assoc-key)
         (map assoc-english)
         (sort-by :is-english)
         first)))

;;- set 1: challenge 7 --------------------------------------------------------

(defn aes-ecb-mode-cipher
  "Use AES-128 in ECB mode to encrypt/decrypt the data."
  [cipher-mode key data]
  (let [secret (SecretKeySpec. (byte-array (map byte key)) "AES")
        cipher (Cipher/getInstance "AES/ECB/NoPadding")
        cm (case cipher-mode
              :encrypt Cipher/ENCRYPT_MODE
              :decrypt Cipher/DECRYPT_MODE)]
    (.init cipher cm secret)
    (.doFinal cipher data)))

(defn aes-ecb-mode-decrypt-file
  "The Base64-encoded content in the file has been encrypted via AES-128 in ECB
  mode.  Decrypt it."
  [file-path key]
  (let [encrypted-data (base64file->bytes file-path)]
    (apply str (bytes->ascii (aes-ecb-mode-cipher :decrypt key encrypted-data)))))

;;- set 1: challenge 8 --------------------------------------------------------

(defn detect-aes-in-ecb-mode
  [block-size data]
  (let [split-blocks #(partition block-size %)
        has-duplicates #(not= (count %) (count (distinct %)))]
    (has-duplicates (split-blocks data))))

(defn detect-aes-in-ecb-mode-file
  [file-path]
  (let [encrypted-datas (string/split-lines (slurp file-path))]
    (filter (partial detect-aes-in-ecb-mode 16) encrypted-datas)))


src/cryptopals/utils.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges: Utility Functions
;;-----------------------------------------------------------------------------

(ns cryptopals.utils
  (:require [clojure.string :as string])
  (:import
    [org.apache.commons.codec.binary Base64 Hex]))

;;- utils ---------------------------------------------------------------------

(defn hex->bytes
  "Converts hex string to byte[]."
  ([]
   (fn [rf]
     (fn
       ([] (rf))
       ([result] (rf result))
       ([result input] (reduce rf result (hex->bytes input))))))
  ([^String hex]
   (Hex/decodeHex (.toCharArray hex))))

(defn bytes->hex
  "Converts byte[] to hex string."
  [bs]
  (apply str (Hex/encodeHex bs)))

(defn byte->hex
  "Converts an integer into string hex."
  [i]
  (bytes->hex (byte-array [i])))

(defn byte->int
  "Converts a byte into an integer."
  [b]
  (bit-and 0xff b))

(def byte->ascii
  "Convert byte to ascii."
  (comp
    char
    byte->int))

(def bytes->ascii
  "Convert byte[] to ascii."
  (partial map byte->ascii))

(def bytes->str
  "Convert byte[] to string."
  (comp
    (partial apply str)
    bytes->ascii))

(defn hex->ascii
  "Convert hex string to ascii string."
  [hex]
  (apply str (into []
                   (comp (hex->bytes)
                         (map byte->ascii))
                   [hex])))

(def ->byte
  "Convert type into a byte."
  (comp unchecked-byte int))

(defn map-values [f m]
  (into {} (map (fn [[k v]] {k (f v)})) m))

(defn sqr-err
  "Calculate the sum of the error squared."
  [observed expected]
  (let [diff (- observed expected)]
    ((comp double /) (* diff diff) expected)))

(defn zip
  "Zip up one or more lists. Note the shortest list will dictate length."
  [& args]
  (apply map vector args))

(defn strip-newlines [text]
  (string/replace text #"\n" ""))

(defn base64file->bytes [file-path]
  (Base64/decodeBase64 ^String (strip-newlines (slurp file-path))))

(defn gcd
  [a b]
  (if (zero? b)
      a
      (recur b, (mod a b))))


src/cryptopals/core.clj
;;-----------------------------------------------------------------------------
;; The Matasano Crypto Challenges
;;-----------------------------------------------------------------------------

(ns cryptopals.core
  (:require [clojure.java.io :as io]
            [clojure.string :as string]
            [cryptopals.set1 :refer [fixed-xor]]
            [cryptopals.utils :refer :all])
  (:import
    [org.apache.commons.codec.binary Base64 Hex]
    [javax.crypto Cipher]
    [javax.crypto.spec SecretKeySpec])
  (:gen-class))

;;- set 2: challenge 9 --------------------------------------------------------

(defn pkcs#7-padding
  "Pad any block to a specific block length, by appending the number of bytes
  of padding to the end of the block."
  [block-length block]
  {:pre [(<= (count block) block-length)
         (<= block-length 256)]}
  (let [pad-size (- block-length (count block))
        padding (take pad-size (repeat (unchecked-byte pad-size)))]
    (into-array Byte/TYPE (concat block padding))))

;;- set 2: challenge 10 -------------------------------------------------------

(defn aes-cbc-mode-decrypt
  "The data has been encrypted via AES-128 in CBC mode.  Decrypt it."
  [key iv data]
  {:pre [(= (count key) (count iv))]}
  (let [secret (SecretKeySpec. (byte-array (map ->byte key)) "AES")
        cipher (Cipher/getInstance "AES/ECB/NoPadding")
        blocks (partition-all (count key) data)
        decrypt (fn [[i c]] (fixed-xor i (.doFinal cipher (byte-array c))))]
    (.init cipher Cipher/DECRYPT_MODE secret)
    (into []
          (comp (map decrypt) cat)
          (zip (cons iv blocks) blocks))))

(defn aes-cbc-mode-decrypt-file
  "The Base64-encoded content in the file has been encrypted via AES-128 in CBC
  mode.  Decrypt it."
  [key iv file-path]
  (let [encrypted-data (base64file->bytes file-path)]
    (apply str (bytes->ascii (aes-cbc-mode-decrypt key iv encrypted-data)))))

(defn aes-cbc-mode-encryptor
  "Encrypts data using AES-128 in CBC mode. Returns a transducer with state,
  tracking the last ciphered data to use for the next xor."
  [key iv]
  {:pre [(= (count key) (count iv))]}
  (fn [rf]
    (let [secret (SecretKeySpec. (byte-array (map (comp unchecked-byte int) key)) "AES")
          cipher (Cipher/getInstance "AES/ECB/NoPadding")
          _ (.init cipher Cipher/ENCRYPT_MODE secret)
          iv-a (atom iv)]
      (fn
        ([] (rf))
        ([result] (rf result))
        ([result input]
         (swap! iv-a #(.doFinal cipher (byte-array (fixed-xor input %))))
         (rf result @iv-a))))))

(defn aes-cbc-mode-encrypt
  "Encrypted the data via AES-128 in CBC mode."
  [key iv data]
  {:pre [(= (count key) (count iv))]}
  (let [xf (comp
             (map (partial pkcs#7-padding (count key)))
             (aes-cbc-mode-encryptor key iv)
             cat)]
    (into [] xf (partition-all (count key) data))))

;;- set 2: challenge 11 -------------------------------------------------------

(defn aes-ecb-mode-encrypt
  "Encrypted the data via AES-128 in ECB mode."
  [key data]
  (let [secret (SecretKeySpec. (byte-array (map ->byte key)) "AES")
        cipher (Cipher/getInstance "AES/ECB/NoPadding")
        encrypt #(.doFinal cipher %)
        xf (comp
             (map (partial pkcs#7-padding (count key)))
             (map encrypt)
             cat)]
    (.init cipher Cipher/ENCRYPT_MODE secret)
    (into [] xf (partition-all (count key) data))))

(defn encryption-oracle [input]
  "Generates a random key and encrypts under it, randomly padding the data
  and picking a encryption mode (ecb/cbc)."
  (letfn [(take-rand [n]
           (take n (repeatedly #(rand-int 256))))]
    (let [key (take-rand 16)
          iv (take-rand 16)
          left-pad (take-rand (+ 5 (rand-int 6)))
          right-pad (take-rand (+ 5 (rand-int 6)))
          xf (comp cat
                   (map ->byte))
          data (into [] xf [left-pad input right-pad])
          mode ([:ecb :cbc] (rand-int 2))]
      (if (= mode :ecb)
        [:ecb (aes-ecb-mode-encrypt key data)]
        [:cbc (aes-cbc-mode-encrypt key iv data)]))))

;;- set 2: challenge 12 -------------------------------------------------------

(defn make-oracle-decryptor [unknown]
  "Encrypts buffers under ECB mode using a consistent but unknown key and pad
  the end with unknown bytes."
  (letfn [(take-rand [n]
            (take n (repeatedly #(rand-int 256))))]
    (let [key (take-rand 16)]
      (fn [input]
        (let [xf (comp cat (map ->byte))
              data (into [] xf [input unknown])]
      (aes-ecb-mode-encrypt key data))))))

(defn discover-block-size
  "Discover the block size of the cipher.  Create a bunch of encryptions and
  find the common divisor, will be the block size.  Since encoding requires
  fixed block size."
  [oracle]
  (let [make-input (fn [a n] (take n (repeat a)))
        input (partial make-input \A)
        encrypt #(oracle (input %))]
    (reduce gcd (map (comp count encrypt) (range 1 40)))))

(defn make-dictionary
  [oracle prefix]
  (let [all-bytes (map ->byte (range 256))
        first-block #(take 16 %)
        xf (comp (map (partial conj (vec prefix)))
                 (map oracle)
                 (map first-block))]
    (zipmap (into [] xf all-bytes)
            all-bytes)))

(defn byte-at-a-time-ecb-decryption [oracle block-size]
  "Create a transducer, return the found byte and keep state of the last found
  bytes to test against.
  1) Knowing the block size, craft an input block that is exactly 1 byte short.
     (for instance, if the block size is 8 bytes, make 'AAAAAAA').
  2) Make a dictionary of every possible last byte by feeding different strings
     to the oracle; for instance, 'AAAAAAAA', 'AAAAAAAB', 'AAAAAAAC',
     remembering the first block of each invocation.
  3) Match the output of the one-byte-short input to one of the entries in your
     dictionary. You've now discovered the first byte of unknown-string."
  (fn [rf]
    (letfn [(make-pad [n]
              (take n (repeat \A)))
            (get-block [b d]
              (drop (* b block-size) (take (* block-size (+ 1 b)) d)))]
      (let [tester (atom (vec (make-pad (- block-size 1))))]
        (fn
          ([] (rf))
          ([result] (rf result))
          ([result input]
           (let [b (quot input block-size)
                 n (rem input block-size)
                 pad-size (- block-size n 1)
                 dict (make-dictionary oracle @tester)
                 block (get-block b (oracle (make-pad pad-size)))
                 found (dict block)]
             (swap! tester #(let [[x & xs] %] (conj (vec xs) found)))
             (rf result found))))))))

;;-----------------------------------------------------------------------------

(defn -main
  ""
  [& args]
  (println "Nothing here."))
